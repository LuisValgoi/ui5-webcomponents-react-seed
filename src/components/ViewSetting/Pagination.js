import React from 'react';
import PropTypes from 'prop-types';
import { FlexBox } from '@ui5/webcomponents-react/lib/FlexBox';
import { FlexBoxAlignItems } from '@ui5/webcomponents-react/lib/FlexBoxAlignItems';
import { Button } from '@ui5/webcomponents-react/lib/Button';
import { ButtonDesign } from '@ui5/webcomponents-react/lib/ButtonDesign';
import { FlexBoxJustifyContent } from '@ui5/webcomponents-react/lib/FlexBoxJustifyContent';
import { spacing } from '@ui5/webcomponents-react-base';

export const Pagination = ({ shouldDisableAll, numberOfElements, totalPages, selectedPage, setPage, style = spacing.sapUiMediumMargin }) => {
  if (!numberOfElements) return null;

  let disablePrevious = false;
  if (selectedPage === 0) {
    disablePrevious = true;
  }
  let disableNext = false;
  if (selectedPage === totalPages - 1) {
    disableNext = true;
  }
  const previousPage = selectedPage - 1;
  const morePreviousPage = previousPage - 1;
  const nextPage = selectedPage + 1;
  const moreNextPage = nextPage + 1;
  const hasPreviousPage = selectedPage > 0;
  const hasMorePreviousPages = selectedPage > 1;
  const hasNextPage = selectedPage < totalPages - 1;
  const hasMoreNextPages = selectedPage < totalPages - 2;

  function setPreviousPage() {
    if (!disablePrevious) setPage(selectedPage - 1);
  }

  function setNextPage() {
    if (!disableNext) setPage(selectedPage + 1);
  }

  return (
    <FlexBox alignItems={FlexBoxAlignItems.Start} style={style} justifyContent={FlexBoxJustifyContent.Center}>
      <Button disabled={disablePrevious || shouldDisableAll} design={ButtonDesign.Transparent} onClick={setPreviousPage}>
        {'<'}
      </Button>

      {hasMorePreviousPages && (
        <Button disabled={shouldDisableAll} design={ButtonDesign.Transparent} style={spacing.sapUiTinyMarginEnd} onClick={() => setPage(morePreviousPage)}>
          {morePreviousPage + 1}
        </Button>
      )}

      {hasPreviousPage && (
        <Button disabled={shouldDisableAll} design={ButtonDesign.Transparent} style={spacing.sapUiTinyMarginEnd} onClick={() => setPage(previousPage)}>
          {previousPage + 1}
        </Button>
      )}

      <Button disabled={shouldDisableAll} style={spacing.sapUiTinyMarginEnd} design={ButtonDesign.Emphasized}>
        {selectedPage + 1}
      </Button>

      {hasNextPage && (
        <Button disabled={shouldDisableAll} design={ButtonDesign.Transparent} style={spacing.sapUiTinyMarginEnd} onClick={() => setPage(nextPage)}>
          {nextPage + 1}
        </Button>
      )}

      {hasMoreNextPages && (
        <Button disabled={shouldDisableAll} design={ButtonDesign.Transparent} style={spacing.sapUiTinyMarginEnd} onClick={() => setPage(moreNextPage)}>
          {moreNextPage + 1}
        </Button>
      )}

      <Button disabled={disableNext || shouldDisableAll} design={ButtonDesign.Transparent} onClick={setNextPage}>
        {'>'}
      </Button>
    </FlexBox>
  );
};

Pagination.propTypes = {
  totalItems: PropTypes.number,
  totalPages: PropTypes.number,
  selectedPage: PropTypes.number,
  setPage: PropTypes.func,
};
