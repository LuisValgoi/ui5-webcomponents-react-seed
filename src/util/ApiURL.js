export default {
  GET_USER_LOGGED: '/v1/user/logged',
  GET_TODO_LIST: '/v1/todo/all',
  GET_TODO_BY_ID: '/v1/todo/:id'
};
