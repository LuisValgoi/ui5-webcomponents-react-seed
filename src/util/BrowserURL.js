export default {
  HOME: '/',
  TODO_LIST: '/todo/list',
  TODO_DETAIL: '/todo/:id/detail',
  NOT_FOUND: '/notFound',
  BUGGY: '/buggy',
  ANY: '/*'
};
