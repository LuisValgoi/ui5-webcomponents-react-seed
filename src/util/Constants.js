export default {
  REACT_QUERY: {
    CODES: {
      SUCCESS: 'success',
      LOADING: 'loading',
      ERROR: 'error',
    },
    KEYS: {
      RQ_GET_USER_LOGGED: 'RQ_GET_USER_LOGGED',
      RQ_GET_TODO_LIST: 'RQ_GET_TODO_LIST',
      GET_TODO_BY_ID: 'GET_TODO_BY_ID',
    },
  },
  SEED: {
    SELECTED_THEME: 'SELECTED_THEME',
  },
};
