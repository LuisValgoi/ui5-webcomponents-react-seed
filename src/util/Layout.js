import React from 'react';
import { Text } from '@ui5/webcomponents-react/lib/Text';

export const TextWrapper = ({ children, isOnHeader, style }) => {
  return (
    <Text
      tooltip={children}
      wrapping
      className={isOnHeader && 'text-header-color'}
      style={{ ...style, lineHeight: '20px' }}
    >
      {children}
    </Text>
  );
};
