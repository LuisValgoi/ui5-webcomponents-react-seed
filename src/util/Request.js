import axios from 'axios';

const Request = axios.create({
  baseURL: process.env.REACT_APP_API_URL
});

export default Request;
