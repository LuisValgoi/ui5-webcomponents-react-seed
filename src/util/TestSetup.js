import React from 'react';

import { render as rtlRender } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';
import { Router } from 'react-router-dom';
import { createMemoryHistory } from 'history';
import { rest } from 'msw';
import { setupServer } from 'msw/node';

import BrowserProvider from './browser/BrowserProvider';

const render = (ui, { route = BrowserProvider.getUrl('HOME'), ...renderOptions } = {}) => {
  const WrapperProvider = ({ ...props }) => {
    const history = createMemoryHistory({ initialEntries: [route] });
    return (
      <BrowserRouter>
        <Router history={history}>{props.children}</Router>
      </BrowserRouter>
    );
  };

  return rtlRender(ui, { wrapper: WrapperProvider, ...renderOptions });
};

const serverCustom = (apiUrl, data) => {
  return setupServer(
    rest.get(`*${apiUrl}`, (req, res, ctx) => {
      return res(ctx.json(data));
    }),
  );
};

export * from '@testing-library/react';

export { render, serverCustom };
