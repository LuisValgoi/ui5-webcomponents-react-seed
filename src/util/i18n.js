import i18n from 'i18next';
import { initReactI18next } from 'react-i18next';

import LanguageDetector from 'i18next-browser-languagedetector';

import translationEN from '../locales/en/translation.json';
import translationPT from '../locales/pt/translation.json';

i18n
  .use(LanguageDetector)
  .use(initReactI18next)
  .init({
    debug: true,
    keySeparator: false,
    interpolation: {
      escapeValue: false,
    },
    resources: {
      en: {
        translation: translationEN,
      },
      pt: {
        translation: translationPT,
      },
    },
    react: {
      useSuspense: false,
    },
  });

export default i18n;
