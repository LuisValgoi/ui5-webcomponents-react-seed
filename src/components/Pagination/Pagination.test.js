import React from 'react';

import '@testing-library/jest-dom/extend-expect';
import { render, screen } from '../../util/TestSetup';

import { Pagination } from './Pagination';

const fourButtonsPagination = {
  numberOfElements: 11,
  totalPages: 2,
  number: 0,
};

describe('Pagination.js Test Suite', () => {
  beforeEach(() => {
    render(<Pagination numberOfElements={fourButtonsPagination.numberOfElements} totalPages={fourButtonsPagination.totalPages} selectedPage={fourButtonsPagination.number} />);
  });

  test('should match snapshot', () => {
    const pagination = screen.getByTestId('pagination-wrapper');

    expect(pagination).toMatchSnapshot();
  });

  test('should have 4 buttons (< 1 2 >) ', () => {
    const pagination = screen.getByTestId('pagination-wrapper');
    const leftarrow = screen.getByTestId('leftarrow-pagination-wrapper');
    const selectedPage = screen.getByTestId('selectedPage-pagination-wrapper');
    const nextPage = screen.getByTestId('nextPage-pagination-wrapper');
    const afterarrow = screen.getByTestId('afterarrow-pagination-wrapper');

    expect(pagination).toBeInTheDocument();
    expect(leftarrow).toHaveTextContent('<');
    expect(selectedPage).toHaveTextContent('1');
    expect(nextPage).toHaveTextContent('2');
    expect(afterarrow).toHaveTextContent('>');
  });

  test('should have button < disabled', () => {
    const pagination = screen.getByTestId('pagination-wrapper');
    const leftarrow = screen.getByTestId('leftarrow-pagination-wrapper');

    expect(pagination).toBeInTheDocument();
    expect(leftarrow).toHaveProperty('disabled', true);
  });

  test('should have button 1 selected', () => {
    const pagination = screen.getByTestId('pagination-wrapper');
    const selectedPage = screen.getByTestId('selectedPage-pagination-wrapper');

    expect(pagination).toBeInTheDocument();
    expect(selectedPage).toHaveProperty('design', 'Emphasized');
  });
});
