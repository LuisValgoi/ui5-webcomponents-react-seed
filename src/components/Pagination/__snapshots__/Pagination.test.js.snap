// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination.js Test Suite should match snapshot 1`] = `
<div
  class="FlexBox-flexBox-0-2-1 FlexBox-flexBoxDirectionRow-0-2-10 FlexBox-justifyContentCenter-0-2-5 FlexBox-alignItemsStart-0-2-12 FlexBox-flexWrapNoWrap-0-2-17"
  data-testid="pagination-wrapper"
  style="margin: 2rem;"
>
  <ui5-button
    data-testid="leftarrow-pagination-wrapper"
    design="Transparent"
    disabled="true"
  >
    &lt;
  </ui5-button>
  <ui5-button
    data-testid="selectedPage-pagination-wrapper"
    design="Emphasized"
    style="margin-right: 0.5rem;"
  >
    1
  </ui5-button>
  <ui5-button
    data-testid="nextPage-pagination-wrapper"
    design="Transparent"
    style="margin-right: 0.5rem;"
  >
    2
  </ui5-button>
  <ui5-button
    data-testid="afterarrow-pagination-wrapper"
    design="Transparent"
  >
    &gt;
  </ui5-button>
</div>
`;
